<a href="https://colab.research.google.com/github/MHamza-Ahmad/Flyrank-Internship/blob/main/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!git clone https://github.com/flyrank-bih/flyrank-ml-internship-starter.git

Cloning into 'flyrank-ml-internship-starter'...
remote: Enumerating objects: 283, done.
remote: Counting objects: 100% (145/145), done.
remote: Compressing objects: 100% (62/62), done.
remote: Total 283 (delta 112), reused 83 (delta 83), pack-reused 138 (from 1)
Receiving objects: 100% (283/283), 1.85 MiB | 4.99 MiB/s, done.
Resolving deltas: 100% (153/153), done.


# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I choose 'Sales Forecasting' for my research lane because it’s where data analytics actually meets practical business strategy. Getting sales predictions right is basically the backbone of any strong organization,it dictates how you handle inventory, manage resources, and keep cash flowing smoothly. With how unpredictable the market is right now, looking into advanced methods like machine learning or time-series analysis is a great way to pull real, actionable insights from messy data and turn past trends into a genuine competitive edge.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

My work helps businesses decide exactly how much product to make, how many supplies to buy, and where to spend their money.

These decisions are used by managers, finance teams, and supply chain workers.

The cost of a wrong guess is high: if the forecast is too high, the company wastes money making and storing items that won't sell. If the forecast is too low, the company runs out of stock, loses money from missed sales, and upsets its customers.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
import pandas as pd

# The starter CSV is located at 'flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv'
try:
    df = pd.read_csv('flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv')
    print("Dataset loaded successfully.")

    # Display the first 5 rows to show some real numbers and data structure
    print("\nFirst 5 rows of the dataset:")
    display(df.head())

    # Display concise summary of a DataFrame including index dtype and column dtypes, non-null values and memory usage.
    print("\nDataFrame Information (2-3 real numbers like total entries, columns, and data types):")
    df.info()

    # Display descriptive statistics for numerical columns
    print("\nDescriptive Statistics for numerical columns:")
    display(df.describe())

except FileNotFoundError:
    print("Error: The specified CSV file was not found. Please check the path.")
except Exception as e:
    print(f"An error occurred while loading the data: {e}")

Dataset loaded successfully.

First 5 rows of the dataset:


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7



DataFrame Information (2-3 real numbers like total entries, columns, and data types):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clic

,search_volume,competition,cpc,word_count,char_count,impressions_90d,clicks_90d,pageviews_90d,sessions_90d,users_90d,...,sessions_prev_30d,content_age_days,age_tier_order,days_since_last_update,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,trend_pct
count,27532.000000,27532.000000,27532.000000,22301.000000,22301.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,...,30000.000000,30000.00000,30000.000000,30000.000000,30000.000000,30000.00000,30000.000000,29875.000000,30000.000000,26612.000000
mean,158.882391,0.146954,0.485342,3107.760325,20665.277835,5200.366300,16.097333,49.942467,37.066633,35.937700,...,10.283000,256.16780,4.786533,46.098300,0.510733,16.34238,2.534520,18.212921,0.768196,-4.785969
std,1518.270825,0.285241,2.101560,1452.382598,10115.344042,16838.019547,75.076958,152.101430,107.069131,103.748185,...,42.578003,132.70793,0.790392,42.078709,3.279162,15.21679,8.310096,29.472768,7.429454,473.861780
min,0.000000,0.000000,0.000000,8.000000,40.000000,1.000000,0.000000,0.000000,1.000000,1.000000,...,0.000000,90.00000,3.000000,1.000000,0.000000,0.00000,0.000000,0.000000,0.000000,-100.000000
25%,0.000000,0.000000,0.000000,2413.000000,15644.000000,81.000000,0.000000,2.000000,2.000000,2.000000,...,1.000000,132.00000,4.000000,20.000000,0.000000,6.20000,0.000000,0.000000,0.000000,-62.600000
50%,10.000000,0.000000,0.000000,2877.000000,19116.000000,731.000000,1.000000,8.000000,7.000000,7.000000,...,2.000000,236.00000,5.000000,20.000000,0.070000,10.80000,0.000000,5.000000,0.000000,-33.500000
75%,20.000000,0.130000,0.000000,3666.000000,24011.000000,3615.250000,7.000000,33.000000,27.000000,27.000000,...,7.000000,333.00000,5.000000,104.000000,0.290000,22.30000,1.350000,23.530000,0.000000,0.000000
max,74000.000000,1.000000,100.360000,9546.000000,111158.000000,517715.000000,4178.000000,5998.000000,4345.000000,4913.000000,...,4247.000000,564.00000,6.000000,373.000000,100.000000,245.00000,100.000000,300.000000,300.000000,44900.000000


For your 'Sales Forecasting' research, here are 3 key metrics from the `content_refresh_anonymized.csv` dataset that stand out:

1.  **`impressions_90d`**: This column represents the total number of impressions a content item received over the last 90 days. With an average of **5200 impressions** (mean from `df.describe()`), this metric is crucial as it indicates the visibility and potential reach of content, directly influencing the likelihood of user interaction and, subsequently, sales.
2.  **`clicks_90d`**: This metric shows the total clicks over the last 90 days, which is a direct measure of user interest and engagement. An average of **16 clicks** per content item (mean) suggests conversion potential. Higher clicks often correlate with higher interest, which can be a leading indicator for sales.
3.  **`sessions_90d`**: Representing the total user sessions over the last 90 days, this metric indicates how often users interacted with the content. With an average of **37 sessions** (mean), this reflects sustained user engagement, which is vital for understanding content performance and its indirect impact on sales. The variability (standard deviation of 107) also suggests some content items drive significantly more sessions than others.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*What I can claim:
My work will show clear patterns based on past data (what we observed) and point out the general direction sales are likely heading. It is designed to act as a reliable guide to help managers make smarter, data-backed choices (decision-support).

What I can never claim:
My work will never offer absolute proof of why something happened (causal proof). It can show that two trends happen at the same time, but it cannot prove that one perfectly caused the other. It will also never act as a crystal ball that guarantees exact future numbers with 100% certainty, nor can it predict unpredictable, massive market shifts (like trying to "predict Google")

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.